src/s2/util/math/exactfloat/bignum.h

```diff
@@ -0,0 +1,457 @@
+    sign_ = -1;
+    return *this;
+  }
+
+  // Unconditionally makes the sign of this bignum positive.
+  Bignum& SetPositive() {
+    sign_ = +1;
+    return *this;
+  }
+
+  // Unconditionally set the sign of this bignum to match the sign of the
+  // argument. If the argument is zero, set the bignum to zero.
+  Bignum& SetSign(int sign) {
+    if (sign == 0) {
+      return SetZero();
+    }
+
+    if (sign < 0) {
+      return SetNegative();
+    }
+    return SetPositive();
+  }
+
+  // Returns true if the number is zero.
+  bool is_zero() const {  //
+    return sign_ == 0;
+  }
+
+  // Returns true if the number is greater than zero.
+  bool positive() const {  //
+    return sign_ > 0;
+  }
+
+  // Returns true if the number is less than zero.
+  bool negative() const {  //
+    return sign_ < 0;
+  }
+
+  // Returns true if the number is odd (least significant bit is 1).
+  bool is_odd() const { return Bit(0); }
+
+  // Returns true if the number is even (least significant bit is 0).
+  bool is_even() const { return !is_odd(); }
+
+  //--------------------------------------
+  // Comparisons.
+  //--------------------------------------
+
+  bool operator==(const Bignum& b) const {
+    return sign_ == b.sign_ && bigits_ == b.bigits_;
+  }
+
+  bool operator!=(const Bignum& b) const { return !(*this == b); }
+  bool operator<(const Bignum& b) const { return Compare(b) < 0; }
+  bool operator<=(const Bignum& b) const { return Compare(b) <= 0; }
+  bool operator>(const Bignum& b) const { return Compare(b) > 0; }
+  bool operator>=(const Bignum& b) const { return Compare(b) >= 0; }
+
+  //--------------------------------------
+  // Arithmetic operators.
+  //--------------------------------------
+
+  Bignum operator+() const { return *this; }
+  Bignum operator-() const;
+
+  // Raise this value to the given power, which must be non-negative.
+  Bignum Pow(int32_t pow) const;
+
+  Bignum& operator+=(const Bignum& b);
+  Bignum& operator-=(const Bignum& b);
+  Bignum& operator*=(const Bignum& b);
+  Bignum& operator<<=(int nbit);
+  Bignum& operator>>=(int nbit);
+
+  friend Bignum operator+(Bignum a, const Bignum& b) { return a += b; }
+  friend Bignum operator-(Bignum a, const Bignum& b) { return a -= b; }
+  friend Bignum operator*(Bignum a, const Bignum& b) { return a *= b; }
+  friend Bignum operator<<(Bignum a, int nbit) { return a <<= nbit; }
+  friend Bignum operator>>(Bignum a, int nbit) { return a >>= nbit; }
+
+ private:
+  // Constructs a Bignum from bigits and an optional sign bit.
+  explicit Bignum(BigitVector bigits, int sign = +1)
+      : bigits_(std::move(bigits)) {
+    NormalizeSign(sign);
+  }
+
+  // Returns the number of bigits in this bignum.
+  size_t size() const {  //
+    return bigits_.size();
+  }
+
+  // Returns true if this value has no digits.
+  bool empty() const {  //
+    return bigits_.empty();
+  }
+
+  // Compare to another bignum, returns -1, 0, +1.
+  int Compare(const Bignum& b) const {
```

Comment:
Seems like a few of these functions are big enough to either go in the `.cc` file or in the implementation section below.